key-rotator/storage/manifest.go

```diff
@@ -0,0 +1,265 @@
+	return path.Join(m.keyPrefix, fmt.Sprintf("%s-manifest.json", dataShareProcessorName))
+}
+
+// datastore represents a given key/value object store backing a Manifest. It
+// includes functionality for getting & putting individual objects by key,
+// specialized for small objects (i.e. no streaming support).
+type datastore interface {
+	// get gets the content of a given key, or returns an error if it can't.
+	// If the key does not exist, an error wrapping ErrObjectNotExist is
+	// returned.
+	get(ctx context.Context, key string) ([]byte, error)
+
+	// put puts the given content to the given key, or returns an error if it
+	// can't.
+	put(ctx context.Context, key string, data []byte) error
+}
+
+type gcsDatastore struct {
+	gcs    *storage.Client
+	bucket string
+}
+
+var _ datastore = gcsDatastore{} // verify gcsDatastore satsifies datastore.
+
+func (ds gcsDatastore) get(ctx context.Context, key string) ([]byte, error) {
+	r, err := ds.gcs.Bucket(ds.bucket).Object(key).NewReader(ctx)
+	if err != nil {
+		if err == storage.ErrObjectNotExist {
+			err = ErrObjectNotExist
+		}
+		return nil, fmt.Errorf("couldn't retrieve gs://%s/%s: %w", ds.bucket, key, err)
+	}
+	defer r.Close() // errchecked Close is below
+	objBytes, err := io.ReadAll(r)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't read gs://%s/%s: %w", ds.bucket, key, err)
+	}
+	if err := r.Close(); err != nil {
+		return nil, fmt.Errorf("couldn't close gs://%s/%s: %w", ds.bucket, key, err)
+	}
+	return objBytes, nil
+}
+
+func (ds gcsDatastore) put(ctx context.Context, key string, data []byte) error {
+	// Canceling a write requires canceling the context, rather than calling
+	// Close(). We therefore create a context we can cancel without affecting
+	// anything else to ensure we don't leave a pending write around in case of
+	// failure.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	w := ds.gcs.Bucket(ds.bucket).Object(key).NewWriter(ctx)
+	w.CacheControl = "no-cache"
+	w.ContentType = "application/json; charset=UTF-8"
+
+	if _, err := w.Write(data); err != nil {
+		return fmt.Errorf("couldn't write gs://%s/%s: %w", ds.bucket, key, err)
+	}
+	if err := w.Close(); err != nil {
+		return fmt.Errorf("couldn't close gs://%s/%s: %w", ds.bucket, key, err)
+	}
+	return nil
+}
+
+type s3Datastore struct {
+	s3     *s3.S3
+	bucket string
+}
+
+var _ datastore = s3Datastore{} // verify s3Datastore satisfies datastore.
+
+func (ds s3Datastore) get(ctx context.Context, key string) ([]byte, error) {
+	objOut, err := ds.s3.GetObjectWithContext(ctx, &s3.GetObjectInput{
+		Bucket: aws.String(ds.bucket),
+		Key:    aws.String(key),
+	})
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == s3.ErrCodeNoSuchKey {
+			err = ErrObjectNotExist
+		}
+		return nil, fmt.Errorf("couldn't retrieve s3://%s/%s: %w", ds.bucket, key, err)
+	}
+	r := objOut.Body
+	defer r.Close() // errchecked Close is below
```

Comment:
As in the GCS case, I'm not certain if it's safe to `Close()` twice, and also I don't think we need to be as fussy about checking on `Close()` on the read path.